pod/src/list.rs

```diff
@@ -0,0 +1,428 @@
+        // Split the buffer to get the length prefix.
+        // buf: [ L L L L | P P D D D D D D D D ...]
+        //       <-------> <---------------------->
+        //       len_bytes          tail
+        let length_size = size_of::<L>();
+        if buf.len() < length_size {
+            return Err(PodSliceError::BufferTooSmall.into());
+        }
+        let (len_bytes, tail) = buf.split_at_mut(length_size);
+
+        // Skip alignment padding to find the start of the data.
+        // tail: [P P | D D D D D D D D ...]
+        //        <-> <------------------->
+        //      padding    data_bytes
+        let padding = calculate_padding::<L, T>()?;
+        let data_bytes = tail
+            .get_mut(padding..)
+            .ok_or(PodSliceError::BufferTooSmall)?;
+
+        // Cast the bytes to typed data
+        let length = pod_from_bytes_mut::<L>(len_bytes)?;
+        let data = pod_slice_from_bytes_mut::<T>(data_bytes)?;
+        let max_length = data.len();
+
+        // Initialize the list or validate its current length.
+        if init {
+            *length = L::try_from(0)?;
+        } else if (*length).into() > max_length {
+            return Err(PodSliceError::BufferTooSmall.into());
+        }
+
+        Ok(Self {
+            length,
+            data,
+            max_length,
+        })
+    }
+
+    /// Unpack the mutable buffer into a mutable slice
+    pub fn unpack<'a>(data: &'a mut [u8]) -> Result<Self, ProgramError>
+    where
+        'a: 'data,
+    {
+        Self::unpack_internal(data, false)
+    }
+
+    /// Unpack the mutable buffer into a mutable slice, and initialize the
+    /// slice to 0-length
+    pub fn init<'a>(data: &'a mut [u8]) -> Result<Self, ProgramError>
+    where
+        'a: 'data,
+    {
+        Self::unpack_internal(data, true)
+    }
+
+    /// Add another item to the slice
+    pub fn push(&mut self, item: T) -> Result<(), ProgramError> {
+        let length = (*self.length).into();
+        if length >= self.max_length {
+            Err(PodSliceError::BufferTooSmall.into())
+        } else {
+            self.data[length] = item;
+            *self.length = L::try_from(length.saturating_add(1))?;
+            Ok(())
+        }
+    }
+
+    /// Remove and return the element at `index`, shifting all later
+    /// elements one position to the left.
+    pub fn remove(&mut self, index: usize) -> Result<T, ProgramError> {
+        let len = (*self.length).into();
+        if index >= len {
+            return Err(ProgramError::InvalidArgument);
+        }
+
+        let removed_item = self.data[index];
+
+        // Move the tail left by one
+        let tail_start = index
+            .checked_add(1)
+            .ok_or(ProgramError::ArithmeticOverflow)?;
+        self.data.copy_within(tail_start..len, index);
+
+        // Zero-fill the now-unused slot at the end
+        let last = len.saturating_sub(1);
+        self.data[last] = T::zeroed();
```

Comment:
nit: up to you, but there's no zeroing during init, so maybe zeroing during removal should be an option? Same as zeroing during init